File: src/slice/reader.rs
//! Slice-based rule file reader.

use fst::Set;
use std::net::IpAddr;

use super::format::*;
use crate::{Error, Result, Target};

/// Reader for slice-based rule files.
pub struct SliceReader {
    data: Vec<u8>,
    header: SliceHeader,
    entries: Vec<SliceEntry>,
}

impl SliceReader {
    /// Load from bytes.
    pub fn from_bytes(data: &[u8]) -> Result<Self> {
        if data.len() < HEADER_SIZE {
            return Err(Error::InvalidHeaderSize {
                expected: HEADER_SIZE,
                actual: data.len(),
            });
        }

        // Parse header
        let header = unsafe { *(data.as_ptr() as *const SliceHeader) };
        header.validate()?;

        // Parse slice entries
        let slice_count = header.slice_count as usize;
        let entry_size = std::mem::size_of::<SliceEntry>();
        let entries_end = HEADER_SIZE + slice_count * entry_size;

        if data.len() < entries_end {
            return Err(Error::Config("Slice index truncated".to_string()));
        }

        let mut entries = Vec::with_capacity(slice_count);
        for i in 0..slice_count {
            let offset = HEADER_SIZE + i * entry_size;
            let entry = unsafe { *(data[offset..].as_ptr() as *const SliceEntry) };
            entries.push(entry);
        }

        Ok(Self {
            data: data.to_vec(),
            header,
            entries,
        })
    }

    /// Get the fallback target.
    pub fn fallback(&self) -> Target {
        self.header.fallback()
    }

    /// Get number of slices.
    pub fn slice_count(&self) -> usize {
        self.entries.len()
    }

    /// Match a domain against all slices in order.
    ///
    /// Returns the target of the first matching slice, or None if no match.
    /// The caller should use `fallback()` when None is returned.
    pub fn match_domain(&self, domain: &str) -> Option<Target> {
        let normalized = domain.to_lowercase();

        for entry in &self.entries {
            if entry.get_type() != Some(SliceType::FstDomain) {
                continue;
            }

            if self.match_domain_in_slice(entry, &normalized) {
                return Some(entry.get_target());
            }
        }

        None
    }

    /// Match an IP address against all slices in order.
    ///
    /// Returns the target of the first matching slice, or None if no match.
    pub fn match_ip(&self, ip: IpAddr) -> Option<Target> {
        for entry in &self.entries {
            match (entry.get_type(), &ip) {
                (Some(SliceType::CidrV4), IpAddr::V4(v4)) => {
                    if self.match_cidr_v4_in_slice(entry, (*v4).into()) {
                        return Some(entry.get_target());
                    }
                }
                (Some(SliceType::CidrV6), IpAddr::V6(v6)) => {
                    if self.match_cidr_v6_in_slice(entry, v6.octets()) {
                        return Some(entry.get_target());
                    }
                }
                _ => continue,
            }
        }

        None
    }

    /// Match a GeoIP country code against all slices in order.
    pub fn match_geoip(&self, country: &str) -> Option<Target> {
        let country_upper = country.to_uppercase();
        let country_bytes = country_upper.as_bytes();

        for entry in &self.entries {
            if entry.get_type() != Some(SliceType::GeoIp) {
                continue;
            }

            if self.match_geoip_in_slice(entry, country_bytes) {
                return Some(entry.get_target());
            }
        }

        None
    }

    fn match_domain_in_slice(&self, entry: &SliceEntry, domain: &str) -> bool {
        let offset = entry.offset as usize;
        let size = entry.size as usize;

        if offset + size > self.data.len() {
            return false;
        }

        let fst_data = &self.data[offset..offset + size];
        let fst = match Set::new(fst_data.to_vec()) {
            Ok(f) => f,
            Err(_) => return false,
        };

        // Try exact match (reversed with dot prefix)
        let with_dot = format!(".{}", domain);
        let reversed: String = with_dot.chars().rev().collect();
        if fst.contains(&reversed) {
            return true;
        }

        // Check parent domains for suffix match
        let parts: Vec<&str> = domain.split('.').collect();
        for i in 1..parts.len() {
            let suffix = parts[i..].join(".");
            let suffix_with_dot = format!(".{}", suffix);
            let reversed_suffix: String = suffix_with_dot.chars().rev().collect();
            if fst.contains(&reversed_suffix) {
                return true;
            }
        }

        false
    }

    fn match_cidr_v4_in_slice(&self, entry: &SliceEntry, ip: u32) -> bool {
        let offset = entry.offset as usize;
        let count = entry.count as usize;

        // Each entry is 8 bytes: network (4) + prefix_len (1) + padding (3)
        for i in 0..count {
            let entry_offset = offset + i * 8;
            if entry_offset + 8 > self.data.len() {
                break;
            }

            let network =
                u32::from_be_bytes(self.data[entry_offset..entry_offset + 4].try_into().unwrap());
            let prefix_len = self.data[entry_offset + 4];

            let mask = if prefix_len == 0 {
                0
            } else if prefix_len >= 32 {
                !0u32
            } else {
                !0u32 << (32 - prefix_len)
            };

            if (ip & mask) == (network & mask) {
                return true;
            }
        }

        false
    }

    fn match_cidr_v6_in_slice(&self, entry: &SliceEntry, ip: [u8; 16]) -> bool {
        let offset = entry.offset as usize;
        let count = entry.count as usize;

        // Each entry is 24 bytes: network (16) + prefix_len (1) + padding (7)
        for i in 0..count {
            let entry_offset = offset + i * 24;
            if entry_offset + 24 > self.data.len() {
                break;
            }

            let network: [u8; 16] = self.data[entry_offset..entry_offset + 16]
                .try_into()
                .unwrap();
            let prefix_len = self.data[entry_offset + 16];

            if self.matches_ipv6_cidr(&ip, &network, prefix_len) {
                return true;
            }
        }

        false
    }

    fn matches_ipv6_cidr(&self, ip: &[u8; 16], network: &[u8; 16], prefix_len: u8) -> bool {
        let full_bytes = prefix_len as usize / 8;
        let remaining_bits = prefix_len as usize % 8;

        if full_bytes > 0 && ip[..full_bytes] != network[..full_bytes] {
            return false;
        }

        if remaining_bits > 0 && full_bytes < 16 {
            let mask = 0xFFu8 << (8 - remaining_bits);
            if (ip[full_bytes] & mask) != (network[full_bytes] & mask) {
                return false;
            }
        }

        true
    }

    fn match_geoip_in_slice(&self, entry: &SliceEntry, country: &[u8]) -> bool {
        let offset = entry.offset as usize;
        let count = entry.count as usize;

        // Each entry is 4 bytes: country_code (2) + padding (2)
        for i in 0..count {
            let entry_offset = offset + i * 4;
            if entry_offset + 4 > self.data.len() {
                break;
            }

            let stored_country = &self.data[entry_offset..entry_offset + 2];
            if country.len() >= 2 && stored_country[0] == country[0] && stored_country[1] == country[1]
            {
                return true;
            }
        }

        false
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::slice::SliceWriter;

    #[test]
    fn test_read_empty_file() {
        let mut writer = SliceWriter::new(Target::Direct);
        let data = writer.build().unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        assert_eq!(reader.fallback(), Target::Direct);
        assert_eq!(reader.slice_count(), 0);
    }

    #[test]
    fn test_match_domain_single_slice() {
        let mut writer = SliceWriter::new(Target::Direct);
        writer
            .add_domain_slice(&["google.com", "youtube.com"], Target::Proxy)
            .unwrap();
        let data = writer.build().unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        assert_eq!(reader.match_domain("google.com"), Some(Target::Proxy));
        assert_eq!(reader.match_domain("www.google.com"), Some(Target::Proxy));
        assert_eq!(reader.match_domain("unknown.com"), None);
    }

    #[test]
    fn test_match_domain_ordering() {
        // Test that slice order determines priority
        let mut writer = SliceWriter::new(Target::Direct);
        // First slice: cn.bing.com -> Direct
        writer
            .add_domain_slice(&["cn.bing.com"], Target::Direct)
            .unwrap();
        // Second slice: bing.com -> Proxy
        writer
            .add_domain_slice(&["bing.com"], Target::Proxy)
            .unwrap();
        let data = writer.build().unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        // cn.bing.com should match first slice (Direct)
        assert_eq!(reader.match_domain("cn.bing.com"), Some(Target::Direct));
        // www.cn.bing.com should also match first slice
        assert_eq!(
            reader.match_domain("www.cn.bing.com"),
            Some(Target::Direct)
        );
        // bing.com should match second slice (Proxy)
        assert_eq!(reader.match_domain("bing.com"), Some(Target::Proxy));
        // www.bing.com should match second slice (Proxy)
        assert_eq!(reader.match_domain("www.bing.com"), Some(Target::Proxy));
    }

    #[test]
    fn test_match_with_fallback() {
        let mut writer = SliceWriter::new(Target::Proxy);
        writer
            .add_domain_slice(&["direct.com"], Target::Direct)
            .unwrap();
        let data = writer.build().unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        // Matched domain
        assert_eq!(reader.match_domain("direct.com"), Some(Target::Direct));
        // Unmatched domain returns None (caller should use fallback)
        assert_eq!(reader.match_domain("unknown.com"), None);
        // Fallback is accessible
        assert_eq!(reader.fallback(), Target::Proxy);
    }

    #[test]
    fn test_match_cidr_v4() {
        let mut writer = SliceWriter::new(Target::Direct);
        // 10.0.0.0/8 -> Direct
        writer
            .add_cidr_v4_slice(&[(0x0A000000, 8)], Target::Direct)
            .unwrap();
        let data = writer.build().unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        assert_eq!(
            reader.match_ip("10.1.2.3".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(reader.match_ip("192.168.1.1".parse().unwrap()), None);
    }

    #[test]
    fn test_match_cidr_v6() {
        let mut writer = SliceWriter::new(Target::Direct);
        // fc00::/7 -> Direct
        let network: [u8; 16] = [0xFC, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0];
        writer
            .add_cidr_v6_slice(&[(network, 7)], Target::Direct)
            .unwrap();
        let data = writer.build().unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        assert_eq!(
            reader.match_ip("fc00::1".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(
            reader.match_ip("fd00::1".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(reader.match_ip("2001:db8::1".parse().unwrap()), None);
    }

    #[test]
    fn test_match_geoip() {
        let mut writer = SliceWriter::new(Target::Proxy);
        writer
            .add_geoip_slice(&["CN", "US"], Target::Direct)
            .unwrap();
        let data = writer.build().unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        assert_eq!(reader.match_geoip("CN"), Some(Target::Direct));
        assert_eq!(reader.match_geoip("US"), Some(Target::Direct));
        assert_eq!(reader.match_geoip("JP"), None);
    }

    #[test]
    fn test_mixed_slice_ordering() {
        let mut writer = SliceWriter::new(Target::Proxy);
        // Order: domain -> cidr -> geoip
        writer
            .add_domain_slice(&["local.lan"], Target::Direct)
            .unwrap();
        writer
            .add_cidr_v4_slice(&[(0x0A000000, 8)], Target::Direct)
            .unwrap();
        writer
            .add_geoip_slice(&["CN"], Target::Direct)
            .unwrap();
        let data = writer.build().unwrap();

        let reader = SliceReader::from_bytes(&data).unwrap();

        assert_eq!(reader.slice_count(), 3);
        assert_eq!(reader.match_domain("local.lan"), Some(Target::Direct));
        assert_eq!(
            reader.match_ip("10.0.0.1".parse().unwrap()),
            Some(Target::Direct)
        );
        assert_eq!(reader.match_geoip("CN"), Some(Target::Direct));
    }
}

File: internal/slice/reader.go
package slice

import (
	"bytes"
	"compress/gzip"
	"fmt"
	"io"
	"net"
	"os"
	"strings"
)

// SliceReader reads and queries K2Rule slice-based rule files
type SliceReader struct {
	data    []byte
	header  *SliceHeader
	entries []*SliceEntry
}

// NewSliceReaderFromBytes loads a SliceReader from raw bytes
func NewSliceReaderFromBytes(data []byte) (*SliceReader, error) {
	if len(data) < HeaderSize {
		return nil, fmt.Errorf("insufficient data for header: got %d bytes, need %d", len(data), HeaderSize)
	}

	// Parse header
	header, err := ParseHeader(data)
	if err != nil {
		return nil, fmt.Errorf("failed to parse header: %w", err)
	}

	if err := header.Validate(); err != nil {
		return nil, fmt.Errorf("invalid header: %w", err)
	}

	// Parse slice entries
	sliceCount := int(header.SliceCount)
	entriesEnd := HeaderSize + sliceCount*EntrySize

	if len(data) < entriesEnd {
		return nil, fmt.Errorf("slice index truncated: expected %d bytes, got %d", entriesEnd, len(data))
	}

	entries := make([]*SliceEntry, 0, sliceCount)
	for i := 0; i < sliceCount; i++ {
		offset := HeaderSize + i*EntrySize
		entry, err := ParseEntry(data[offset:])
		if err != nil {
			return nil, fmt.Errorf("failed to parse entry %d: %w", i, err)
		}
		entries = append(entries, entry)
	}

	return &SliceReader{
		data:    data,
		header:  header,
		entries: entries,
	}, nil
}

// NewSliceReaderFromGzip loads a SliceReader from gzip-compressed bytes
func NewSliceReaderFromGzip(gzipData []byte) (*SliceReader, error) {
	gr, err := gzip.NewReader(bytes.NewReader(gzipData))
	if err != nil {
		return nil, fmt.Errorf("failed to create gzip reader: %w", err)
	}
	defer gr.Close()

	data, err := io.ReadAll(gr)
	if err != nil {
		return nil, fmt.Errorf("failed to decompress gzip data: %w", err)
	}

	return NewSliceReaderFromBytes(data)
}

// NewSliceReaderFromFile loads a SliceReader from a file (auto-detects gzip)
func NewSliceReaderFromFile(path string) (*SliceReader, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("failed to read file: %w", err)
	}

	// Check if it's gzipped (magic bytes: 0x1f 0x8b)
	if len(data) >= 2 && data[0] == 0x1f && data[1] == 0x8b {
		return NewSliceReaderFromGzip(data)
	}

	return NewSliceReaderFromBytes(data)
}

// Fallback returns the fallback target as uint8
func (r *SliceReader) Fallback() uint8 {
	return r.header.Fallback()
}

// SliceCount returns the number of slices
func (r *SliceReader) SliceCount() int {
	return len(r.entries)
}

// MatchDomain matches a domain against all domain slices
// Returns the target of the first matching slice, or nil if no match
func (r *SliceReader) MatchDomain(domain string) *uint8 {
	normalized := strings.ToLower(domain)

	for _, entry := range r.entries {
		if entry.GetType() != SliceTypeFstDomain {
			continue
		}

		if r.matchDomainInSlice(entry, normalized) {
			target := entry.GetTarget()
			return &target
		}
	}

	return nil
}

// MatchIP matches an IP address against all IP slices
// Returns the target of the first matching slice, or nil if no match
func (r *SliceReader) MatchIP(ip net.IP) *uint8 {
	for _, entry := range r.entries {
		switch entry.GetType() {
		case SliceTypeCidrV4:
			if ip4 := ip.To4(); ip4 != nil {
				ipv4 := ipToUint32(ip4)
				if r.matchCidrV4InSlice(entry, ipv4) {
					target := entry.GetTarget()
					return &target
				}
			}
		case SliceTypeCidrV6:
			if ip16 := ip.To16(); ip16 != nil {
				if r.matchCidrV6InSlice(entry, [16]byte(ip16)) {
					target := entry.GetTarget()
					return &target
				}
			}
		}
	}

	return nil
}

// MatchGeoIP matches a GeoIP country code against all GeoIP slices
// Returns the target of the first matching slice, or nil if no match
func (r *SliceReader) MatchGeoIP(country string) *uint8 {
	countryUpper := strings.ToUpper(country)
	countryBytes := []byte(countryUpper)

	for _, entry := range r.entries {
		if entry.GetType() != SliceTypeGeoIP {
			continue
		}

		if r.matchGeoIPInSlice(entry, countryBytes) {
			target := entry.GetTarget()
			return &target
		}
	}

	return nil
}

// matchDomainInSlice matches a domain within a single FST domain slice
func (r *SliceReader) matchDomainInSlice(entry *SliceEntry, domain string) bool {
	offset := int(entry.Offset)
	size := int(entry.Size)

	if offset+size > len(r.data) {
		return false
	}

	fstData := r.data[offset : offset+size]

	// Create FST from data
	fst, err := NewFSTReader(fstData)
	if err != nil {
		return false
	}

	// Try exact match (reversed with dot prefix)
	withDot := "." + domain
	reversed := reverseString(withDot)
	if fst.Contains([]byte(reversed)) {
		return true
	}

	// Check parent domains for suffix match
	parts := strings.Split(domain, ".")
	for i := 1; i < len(parts); i++ {
		suffix := strings.Join(parts[i:], ".")
		suffixWithDot := "." + suffix
		reversedSuffix := reverseString(suffixWithDot)
		if fst.Contains([]byte(reversedSuffix)) {
			return true
		}
	}

	return false
}

// matchCidrV4InSlice matches an IPv4 address within a single CIDR v4 slice
func (r *SliceReader) matchCidrV4InSlice(entry *SliceEntry, ip uint32) bool {
	offset := int(entry.Offset)
	count := int(entry.Count)

	// Each entry is 8 bytes: network (4) + prefix_len (1) + padding (3)
	for i := 0; i < count; i++ {
		entryOffset := offset + i*8
		if entryOffset+8 > len(r.data) {
			break
		}

		// Network is in big-endian (network byte order)
		network := uint32(r.data[entryOffset])<<24 |
			uint32(r.data[entryOffset+1])<<16 |
			uint32(r.data[entryOffset+2])<<8 |
			uint32(r.data[entryOffset+3])

		prefixLen := r.data[entryOffset+4]

		// Calculate mask
		var mask uint32
		if prefixLen == 0 {
			mask = 0
		} else if prefixLen >= 32 {
			mask = ^uint32(0)
		} else {
			mask = ^uint32(0) << (32 - prefixLen)
		}

		if (ip & mask) == (network & mask) {
			return true
		}
	}

	return false
}

// matchCidrV6InSlice matches an IPv6 address within a single CIDR v6 slice
func (r *SliceReader) matchCidrV6InSlice(entry *SliceEntry, ip [16]byte) bool {
	offset := int(entry.Offset)
	count := int(entry.Count)

	// Each entry is 24 bytes: network (16) + prefix_len (1) + padding (7)
	for i := 0; i < count; i++ {
		entryOffset := offset + i*24
		if entryOffset+24 > len(r.data) {
			break
		}

		var network [16]byte
		copy(network[:], r.data[entryOffset:entryOffset+16])
		prefixLen := r.data[entryOffset+16]

		if matchesIPv6CIDR(&ip, &network, prefixLen) {
			return true
		}
	}

	return false
}

// matchGeoIPInSlice matches a country code within a single GeoIP slice
func (r *SliceReader) matchGeoIPInSlice(entry *SliceEntry, country []byte) bool {
	offset := int(entry.Offset)
	count := int(entry.Count)

	// Each entry is 4 bytes: country_code (2) + padding (2)
	for i := 0; i < count; i++ {
		entryOffset := offset + i*4
		if entryOffset+4 > len(r.data) {
			break
		}

		storedCountry := r.data[entryOffset : entryOffset+2]
		if len(country) >= 2 && storedCountry[0] == country[0] && storedCountry[1] == country[1] {
			return true
		}
	}

	return false
}

// Helper functions

func ipToUint32(ip net.IP) uint32 {
	if len(ip) == 16 {
		ip = ip[12:16]
	}
	return uint32(ip[0])<<24 | uint32(ip[1])<<16 | uint32(ip[2])<<8 | uint32(ip[3])
}

func matchesIPv6CIDR(ip, network *[16]byte, prefixLen uint8) bool {
	fullBytes := int(prefixLen / 8)
	remainingBits := int(prefixLen % 8)

	// Compare full bytes
	if fullBytes > 0 {
		for i := 0; i < fullBytes; i++ {
			if ip[i] != network[i] {
				return false
			}
		}
	}

	// Compare remaining bits
	if remainingBits > 0 && fullBytes < 16 {
		mask := uint8(0xFF << (8 - remainingBits))
		if (ip[fullBytes] & mask) != (network[fullBytes] & mask) {
			return false
		}
	}

	return true
}

func reverseString(s string) string {
	runes := []rune(s)
	for i, j := 0, len(runes)-1; i < j; i, j = i+1, j-1 {
		runes[i], runes[j] = runes[j], runes[i]
	}
	return string(runes)
}
